File: server/models/fractal.go
package models

import "fmt"

type Vector2F struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

func (a *Vector2F) Add(b Vector2F) Vector2F {
	return Vector2F{
		X: a.X + b.X,
		Y: a.Y + b.Y,
	}
}

type Vector3F struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
	Z float64 `json:"z"`
}

func (a *Vector3F) Add(b Vector3F) Vector3F {
	return Vector3F{
		X: a.X + b.X,
		Y: a.Y + b.Y,
		Z: a.Z + b.Z,
	}
}

type ColorStep struct {
	Value float64  `json:"value"`
	Color Vector3F `json:"color"`
}

type FractalDimensions struct {
	Width  int `json:"width"`
	Height int `json:"height"`
	MinX   int `json:"view_min_x"`
	MinY   int `json:"view_min_y"`
	MaxX   int `json:"view_max_x"`
	MaxY   int `json:"view_max_y"`
}

type RequestFractalJulia struct {
	Power      int               `json:"power" binding:"required"`
	Constant   Vector2F          `json:"constant" binding:"required"`
	Center     Vector2F          `json:"center" binding:"required"`
	Scale      float64           `json:"scale" binding:"required"`
	Iterations int               `json:"iterations" binding:"required"`
	Samples    int               `json:"samples" binding:"required"`
	Dimensions FractalDimensions `json:"dimensions" binding:"required"`
	Colors     []ColorStep       `json:"colors" binding:"required"`
}

type RequestFractalJuliaVideo struct {
	Start   RequestFractalJulia `json:"start"`
	End     RequestFractalJulia `json:"end"`
	Seconds int                 `json:"seconds"`
}

type RequestFractalMandelbrot struct {
	Constant   Vector2F          `json:"constant" binding:"required"`
	Center     Vector2F          `json:"center" binding:"required"`
	Scale      float64           `json:"scale" binding:"required"`
	Iterations int               `json:"iterations" binding:"required"`
	Samples    int               `json:"samples" binding:"required"`
	Dimensions FractalDimensions `json:"dimensions" binding:"required"`
	Colors     []ColorStep       `json:"colors" binding:"required"`
}

type ResponseFractal struct {
	Url string `json:"url"`
}

type ResponseFractalStartVideo struct {
	Id string `json:"id"`
}

func NewFractalJuliaFilename(r RequestFractalJulia, prefix string, extension string) string {
	if prefix != "" {
		prefix += "_"
	}
	return fmt.Sprintf("%sp%d_kx%f_ky%f_cx%f_cy%f_sc%f_it%d_sa%d_dw%d_dh%d_dmix%d_dmiy%d_dmax%d_dmay%d.%s", prefix, r.Power, r.Constant.X, r.Constant.Y, r.Center.X, r.Center.Y, r.Scale, r.Iterations, r.Samples, r.Dimensions.Width, r.Dimensions.Height, r.Dimensions.MinX, r.Dimensions.MinY, r.Dimensions.MaxX, r.Dimensions.MaxY, extension)
}

File: api/models/fractal.py
from pydantic import BaseModel
import numpy as np

from typing import Optional, List





class Vector2F(BaseModel):
    x: float
    y: float


class Vector3F(BaseModel):
    x: float
    y: float
    z: float


class ColorStep(BaseModel):
    value: float
    color: Vector3F


class FractalDimensions(BaseModel):
    width: int
    height: int
    view_min_x: int
    view_min_y: int
    view_max_x: int
    view_max_y: int


class FractalConfig(BaseModel):
    center: Vector2F
    scale: float
    iterations: int
    samples: int
    dimensions: FractalDimensions
    colors: Optional[List[ColorStep]]

class JuliaConfig(FractalConfig):
    power: int
    constant: Vector2F
    

class MandelbrotConfig(FractalConfig):
    ...

class VideoConfig(BaseModel):
    start: FractalConfig
    end: FractalConfig
    seconds: int

class JuliaVideoConfig(BaseModel):
    start: JuliaConfig
    end: JuliaConfig
    seconds: int

class MandelbrotVideoConfig(BaseModel):
    start: MandelbrotConfig
    end: MandelbrotConfig
    seconds: int


def julia_filename(r: JuliaConfig, prefix: str, extension: str) -> str:
	if prefix != "":
		prefix += "_"
          
	return "julia_{0}p{1}_kx{2}_ky{3}_cx{4}_cy{5}_sc{6}_it{7}_sa{8}_dw{9}_dh{10}_dmix{11}_dmiy{12}_dmax{13}_dmay{14}.{15}".format(prefix, r.power, r.constant.x, r.constant.y, r.center.x, r.center.y, r.scale, r.iterations, r.samples, r.dimensions.width, r.dimensions.height, r.dimensions.view_min_x, r.dimensions.view_min_y, r.dimensions.view_max_x, r.dimensions.view_max_y, extension)


def mandelbrot_filename(r: MandelbrotConfig, prefix: str, extension: str) -> str:
	if prefix != "":
		prefix += "_"
          
	return "mandelbrot_{0}_cx{1}_cy{2}_sc{3}_it{4}_sa{5}_dw{6}_dh{7}_dmix{8}_dmiy{9}_dmax{10}_dmay{11}.{12}".format(prefix, r.center.x, r.center.y, r.scale, r.iterations, r.samples, r.dimensions.width, r.dimensions.height, r.dimensions.view_min_x, r.dimensions.view_min_y, r.dimensions.view_max_x, r.dimensions.view_max_y, extension)